homeassistant/components/systemmonitor/coordinator.py

```diff
@@ -0,0 +1,131 @@
+"""DataUpdateCoordinators for the System monitor integration."""
+from __future__ import annotations
+
+from datetime import timedelta
+from functools import cache
+import logging
+import os
+from typing import Literal
+
+import psutil
+from psutil._common import sdiskusage, shwtemp, snetio, snicaddr, sswap
+from psutil._pslinux import svmem
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_component import DEFAULT_SCAN_INTERVAL
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import DOMAIN
+from .sensor import CPU_SENSOR_PREFIXES
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SystemMonitorDiskCoordinator(
+    DataUpdateCoordinator[
+        (
+            sdiskusage
+            | sswap
+            | svmem
+            | dict[str, snetio]
+            | dict[str, list[snicaddr]]
+            | tuple[float, float, float]
+            | float
+            | None
+        )
+    ]
+):
+    """A System monitor Data Update Coordinator."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        type_: Literal["disk", "swap", "memory", "netio", "netaddr", "load", "cputemp"],
+        argument: str,
+    ) -> None:
+        """Initialize the coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=DEFAULT_SCAN_INTERVAL.seconds),
+        )
+        self._type = type_
+        self._argument = argument
+
+    async def _async_update_data(
+        self,
+    ) -> (
+        sdiskusage
+        | sswap
+        | svmem
+        | dict[str, snetio]
+        | dict[str, list[snicaddr]]
+        | tuple[float, float, float]
+        | float
+        | None
+    ):
+        """Fetch data."""
+
+        if self._type == "disk":
```

Comment:
Create one coordinator class per measurement call type instead. Each measurement will have a different return value that will correspond to the coordinator data type.